src/test/java/org/elasticsearch/index/shard/IndexShardTests.java

```diff
@@ -63,4 +63,96 @@ public void testFlushOnDeleteSetting() throws Exception {
         assertEquals(newValue, shard.isFlushOnClose());
 
     }
+
+    public void testWriteShardState() throws Exception {
+        try (NodeEnvironment env = newNodeEnvironment()) {
+            ShardId id = new ShardId("foo", 1);
+            long version = between(1, Integer.MAX_VALUE / 2);
+            boolean primary = randomBoolean();
+            ShardStateMetaData state1 = new ShardStateMetaData(version, primary, "foo");
+            ShardStateMetaData.write(logger, "foo", id, state1, randomBoolean(), env.shardPaths(id));
+            ShardStateMetaData shardStateMetaData = ShardStateMetaData.load(logger, id, env.shardPaths(id));
+            assertEquals(shardStateMetaData, state1);
+
+            ShardStateMetaData state2 = new ShardStateMetaData(version, primary, "foo");
+            ShardStateMetaData.write(logger, "foo", id, state2, randomBoolean(), env.shardPaths(id));
+            shardStateMetaData = ShardStateMetaData.load(logger, id, env.shardPaths(id));
+            assertEquals(shardStateMetaData, state1);
+
+            ShardStateMetaData state3 = new ShardStateMetaData(version + 1, primary, "foo");
+            ShardStateMetaData.write(logger, "foo", id, state3, randomBoolean(), env.shardPaths(id));
+            shardStateMetaData = ShardStateMetaData.load(logger, id, env.shardPaths(id));
+            assertEquals(shardStateMetaData, state3);
+            assertEquals("foo", state3.indexUUID);
+        }
+    }
+
+    public void testPersistenceStateMetadataPersistence() throws Exception {
+        createIndex("test");
+        ensureGreen();
+        IndicesService indicesService = getInstanceFromNode(IndicesService.class);
+        NodeEnvironment env = getInstanceFromNode(NodeEnvironment.class);
+        IndexService test = indicesService.indexService("test");
+        IndexShard shard = test.shard(0);
+        ShardStateMetaData shardStateMetaData = ShardStateMetaData.load(logger, shard.shardId, env.shardPaths(shard.shardId));
+        assertEquals(getShardStateMetadata(shard), shardStateMetaData);
+        ShardRouting originalRouting = shard.shardRouting;
+        ShardRouting routing = new MutableShardRouting(shard.shardRouting, shard.shardRouting.version()+1);
+        shard.updateRoutingEntry(routing, true);
+
+        shardStateMetaData = ShardStateMetaData.load(logger, shard.shardId, env.shardPaths(shard.shardId));
+        assertEquals(shardStateMetaData, getShardStateMetadata(shard));
```

Comment:
maybe use the routing created in the test when creating the ShardStateMetadata? seems better then using the one in the shard?